File: src/main/java/frc/robot/subsystems/OI.java
package frc.robot.subsystems;

import frc.robot.utils.Range;
import frc.robot.utils.Utils;
import frc.robot.Constants.oiConstants;
import frc.robot.Constants.drivetrainConstants;

import edu.wpi.first.wpilibj.XboxController;
import edu.wpi.first.wpilibj2.command.SubsystemBase;

public class OI extends SubsystemBase{
    XboxController driver_controller = new XboxController(0);
    XboxController operator_controller = new XboxController(1);

    private double deadzone(double i, double dz){
        if (Math.abs(i) <= dz){
            return 0;
        }
        else {
            return i;
        }
    }

    //Driver

    public boolean ready_straight_assist(){
        double[] output = get_raw_output();
        return (output[0] <= oiConstants.drivetrain_rotation_assist_deadband);
    }

    public double[] get_raw_output(){
        double x = driver_controller.getRightX();
        double y = driver_controller.getLeftY();
        return new double[]{
            x,
            y // Moving the controller to the left gives a negative number, but should represent a positive (clockwise) rotation
        };
    }

    public double[] get_curvature_output(){
        double[] output = get_raw_output();
        double x = -Math.copySign(Math.abs(deadzone(output[0], oiConstants.driver_deadband)), output[0]);
        double y = Math.copySign(Math.abs(deadzone(output[1], oiConstants.driver_deadband)), output[1]);
        if (get_beast_mode()){
            y *= -1;
        }
        return new double[]{x*.5, y*.7};
    }

    public boolean get_beast_mode(){
        return driver_controller.getBButton();
    }

    public boolean get_update_pid_dash(){
        return driver_controller.getXButtonPressed();
    }
    
    public boolean get_toggle_pid_type_pressed(){
        return driver_controller.getBackButtonPressed();
    }

    public boolean get_enable_pid(){
        return driver_controller.getAButtonPressed();
    }

    public boolean get_manual_control_override(){  
        return driver_controller.getBButtonPressed();
    }

    public boolean get_update_telemetry(){
        return driver_controller.getYButtonPressed();
    }

    public double get_turbo_mode_modifier(){
        return driver_controller.getRightTriggerAxis();
    }

    public double process_turbo_mode(){
        double modifier = get_turbo_mode_modifier();
        double x = Utils.mapValueInRange(modifier, new Range(0,1), new Range(drivetrainConstants.max_motor_power, drivetrainConstants.turbo_mode_power));
        return x;   
    }

    //Operator

    public boolean get_intake_outtake(){
        return operator_controller.getAButton();
    }

    public boolean get_intake_intake(){
        return operator_controller.getBButton();
    }
        
    public boolean get_intake_raise(){
        return operator_controller.getRightBumperPressed();
    }

    public boolean get_intake_lower(){
        return operator_controller.getLeftBumperPressed();
    }

    public boolean get_shooter_spinup(){
        return operator_controller.getXButton();
    }

    public boolean get_shooter_fire(){
        return operator_controller.getYButton();
    }

    public boolean get_position_control(){
        return operator_controller.getBackButtonPressed();
    }
    
    public boolean get_rotation_control(){
        return operator_controller.getStartButtonPressed();
    }
}

File: src/main/java/frc/robot/subsystems/OI.kt
package frc.robot.subsystems

import edu.wpi.first.wpilibj.XboxController
import edu.wpi.first.wpilibj2.command.SubsystemBase
import frc.robot.utils.Point
import kotlin.math.abs
import kotlin.math.withSign

private const val rotationAssistDeadband = .05
private const val driverDeadband = .05
private const val maxMotorPower = .4
private const val turboModePower = .57

class OI : SubsystemBase() {
    var driver_controller = XboxController(0)
    var operator_controller = XboxController(1)

    private fun deadzone(i: Double, dz: Double): Double {
        return if (abs(i) <= dz) {
            0.0
        } else {
            i
        }
    }

    fun readyStraightAssist(): Boolean {
        val output = getRawOutput()
        return output.x <= rotationAssistDeadband
    }

    fun getRawOutput(): Point {
        val x = driver_controller.rightX
        val y = driver_controller.leftY
        return Point(x, -y)
    }

    fun getCurvatureOutput(): Point {
        val output = getRawOutput()
        val x = -abs(deadzone(output.x, driverDeadband)).withSign(output.x)
        var y = abs(deadzone(output.y, driverDeadband)).withSign(output.y)
        if (getBeastMode()) {
            y *= -1.0
        }
        return Point(x, y)
    }

    fun getBeastMode(): Boolean {
        return driver_controller.bButton
    }

    fun getUpdatePidDash(): Boolean {
        return driver_controller.xButtonPressed
    }

    fun getTogglePidTypePressed(): Boolean {
        return driver_controller.backButtonPressed
    }

    fun getEnablePid(): Boolean {
        return driver_controller.aButtonPressed
    }

    fun getManualControlOverride(): Boolean {
        return driver_controller.bButtonPressed
    }

    fun getUpdateTelemetry(): Boolean {
        return driver_controller.yButtonPressed
    }

    fun getTurboModeModifier(): Double {
        return driver_controller.rightTriggerAxis
    }

    // public double process_turbo_mode(){
    //     double modifier = get_turbo_mode_modifier();
    //     double x = map_value(modifier, 0, 1, max_motor_power, turbo_mode_power);
    //     return x;
    // }
}